torch/csrc/jit/tensorexpr/operators/misc.h

```diff
@@ -42,6 +42,31 @@ ExprHandle scalarOrConstant(const ArgValue& v);
 ExprHandle broadcast(BufHandle b, const std::vector<ExprHandle>& axes);
 ExprHandle constant(const ArgValue& v);
 
+// Infrastructure to provide indirect-indexing related lowering. It helps
+// generating the overall loop-nest for the Op and indirect-indexing related
+// logic, while leaving Op-specific logic to be defined by Op with injected
+// function. E.x.: take a 2D indices and a 3D target (to be indirect indexing),
+// and the 2nd dim is the dim to be indirect-indexing for i : size_i
```

Comment:
Can you explain the usage of the arguments by this example?

Reply:
computeIndirectIndexing() should be called by a NNC lowering function like computeEmbedding().

posOfIndices - position of indices in Inputs of the original lowering function, which should be 1 for embedding
posOfIdxingTarget - position of target (to be indirect indexing) in inputs of the original lowering function, which should be 0 for embedding
dimOfIndirectIdxing - Indirect indexing dim of target, which should be 1 (the 2nd dim) for this case

maybe I should direct passing the indices and tensor NNC BufHandle instead of the position and the entire Inputs, which may be more self-explainable.

Reply:
> nit: formatting

Do you mean lint formatting?

Reply:
> computeIndirectIndexing() should be called by a NNC lowering function like computeEmbedding().
> 
> posOfIndices - position of indices in Inputs of the original lowering function, which should be 1 for embedding posOfIdxingTarget - position of target (to be indirect indexing) in inputs of the original lowering function, which should be 0 for embedding dimOfIndirectIdxing - Indirect indexing dim of target, which should be 1 (the 2nd dim) for this case
> 
> maybe I should direct passing the indices and target NNC BufHandle instead of the positions and the entire Inputs, which may be more self-explainable.

Oh, I mean to explain that in the comment. And explain with this example.

Reply:
> > nit: formatting
> 
> Do you mean lint formatting?

Oh, I mean making `for i : size_t` a new line.

Reply:
> > > nit: formatting
> > 
> > 
> > Do you mean lint formatting?
> 
> Oh, I mean making `for i : size_t` a new line.

Got you. Originally I put `for i : size_t` a new line as to adapt to below for-loops. However, I run lint-runner and it forces changing to this ... I will try to see how to walk around this.

Reply:
> > computeIndirectIndexing() should be called by a NNC lowering function like computeEmbedding().
> > posOfIndices - position of indices in Inputs of the original lowering function, which should be 1 for embedding posOfIdxingTarget - position of target (to be indirect indexing) in inputs of the original lowering function, which should be 0 for embedding dimOfIndirectIdxing - Indirect indexing dim of target, which should be 1 (the 2nd dim) for this case
> > maybe I should direct passing the indices and target NNC BufHandle instead of the positions and the entire Inputs, which may be more self-explainable.
> 
> Oh, I mean to explain that in the comment. And explain with this example.

Sure, I can update the comment to explain more. BTW, do you think it will be more direct and self-explain to pass indices and target NNC BufHandle as params instead of current positions and Inputs?

Reply:
Can you elaborate the idea with a function prototype?

Reply:
> Can you elaborate the idea with a function prototype?

```
StmtPtr computeIndirectIndexing(
    const BufHandle& indices, // indices for indirect-indexing
    const BufHandle& idxingTarget, // target tensor for indirect indexing
    size_t dimOfIndirectIdxing, // Indirect indexing dim of target
    const std::vector<ExprHandle>& outputShape, // Original lowering outputShape
    const std::function<StmtPtr(
        const ExprPtr&, // Loaded IdxingTarget
        const std::vector<ExprPtr>& // Loop-nest Indices
        )>& innerStmtFunc // OP-specific customization for inner-most loop body
                          // generation
);
```

Reply:
Yes, it is clearer. BTW, it is basically a unary point-wise with indirect indexing on one of the dim, right? Is it a too specific abstraction? Do you think we should call out such semantics in the function name?

Reply:
I thought about this previously, but stay on current naming as because ---- from capability perspective, this function can support non point-wise operations, like loop-carried operations: output[i, j, k] = output[i, j, k-1] or even more complicated computation. And also this function can help with Ops with binary or even more operands with the only constrain that only one operand can be indirect-indexing. I cannot come up with a detailed Op for these scenarios, but this function can support this possibility.

Reply:
I don't have strong opinion here but I think designing the semantics with DL use cases in consideration would make things simpler. The use cases in my mind are unary pointwise (embedding lookup, index select, upsamping etc.) and sparse matmul (sparse tensor with CSR mutiplied by a dense tensor). Do you target the interface to cover both scenarios or the former? Is there any other use case you want to target?

Reply:
Currently it is targeting the former, there is no sparse support in NNC yet I am not sure whether this interface can fully support spmm related.

Reply:
OK, that's why I thought making the naming/semantics specific for the pointwise use case would be simpler? Anyway, you decide. :)